front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts

```diff
@@ -74,58 +82,107 @@ const getTrainrunSectionsByTrainrunId = (netzgrafikDto: NetzgrafikDto, trainrunI
+  // each section's source or target to a transition via a port. Example:
+  //
+  //     const node = { id: 10, transitions: [{ port1Id: 30, port2Id: 31 }], … };
+  //     const leftSection = { id: 20, targetNodeId: 10, targetPortId: 30, … };
+  //     const rightSection = { id: 21, sourceNodeId: 10, sourcePortId: 31, … };
+  //
+  // Note, there is no guarantee that a train run will be made up of stricly
+  // ordered source → target chains. Source and target are arbitrary names for
+  // a section's ends. For instance, two sections' sources might point to the
+  // same node:
   //
-  // Two subsequent sections can be linked together at a node with a target
-  // port followed by a transition itself followed by a source port.
+  //     const node = { id: 10, transitions: [{ port1Id: 30, port2Id: 31 }], … };
+  //     const leftSection = { id: 20, sourceNodeId: 10, sourcePortId: 30, … };
+  //     const rightSection = { id: 21, sourceNodeId: 10, sourcePortId: 31, … };
   //
-  // Build a map of sections keyed by their previous section's target port ID.
-  // Find the departure section: it's the one without a transition for its
-  // source port.
-  let departureSection: TrainrunSectionDto | undefined;
-  const sectionsByPrevTargetPortId = new Map<number, TrainrunSectionDto>();
+  // Build a map of sections keyed by the outgoing port ID their target is
+  // connected to. Build another symmetrical map for sources. Find the leaf
+  // (departure/arrival) sections: these are the ones without a transition for
+  // their source or target port.
+  const sectionsByConnectedTargetPortId = new Map<number, TrainrunSectionDto>();
+  const sectionsByConnectedSourcePortId = new Map<number, TrainrunSectionDto>();
+  const leafSectionsAndNodes: [TrainrunSectionDto, number][] = [];
   for (const section of sections) {
     const sourceNode = getNodeById(netzgrafikDto.nodes, section.sourceNodeId)!;
-    const transition = sourceNode.transitions.find(
-      (tr) => tr.port1Id === section.sourcePortId || tr.port2Id === section.sourcePortId
-    );
-    if (transition) {
-      const prevPortId =
-        transition.port1Id === section.sourcePortId ? transition.port2Id : transition.port1Id;
-      sectionsByPrevTargetPortId.set(prevPortId, section);
+    const targetNode = getNodeById(netzgrafikDto.nodes, section.targetNodeId)!;
+
+    const sourceConnectedPortId = findConnectedPortId(sourceNode, section.sourcePortId);
+    const targetConnectedPortId = findConnectedPortId(targetNode, section.targetPortId);
+
+    if (sourceConnectedPortId !== null) {
+      sectionsByConnectedTargetPortId.set(sourceConnectedPortId, section);
     } else {
-      departureSection = section;
+      leafSectionsAndNodes.push([section, section.sourceNodeId]);
+    }
+    if (targetConnectedPortId !== null) {
+      sectionsByConnectedSourcePortId.set(targetConnectedPortId, section);
+    } else {
+      leafSectionsAndNodes.push([section, section.targetNodeId]);
     }
```

Comment:
I noticed that if both `sourceConnectedPortId `and `targetConnectedPortId `are null,  this section gets added twice (once for source, once for target) as a leaf.
Can this lead to unexpected behaviour in the case where there are not connected ?

Reply:
This can happen when a train run has a single section. In that case, the lone section has no neighbor so both of these variables will be null.

This won't lead to unexpected behavior, because in the loop below we check whether we've already seen a section before iterating over a path. So we'll start walking on the section from one end, and then skip the other one.